### ISSUE TITLE
get_class(): Argument #1 ($object) must be of type object, string given in magento/framework/Setup/Patch/PatchApplier

### ISSUE THREAD

**Issue by speller** [magento/magento2#37545]
### Preconditions and environment

- Magento version 2.4.6
- Unknown

### Steps to reproduce

Run `setup:upgrade`

### Expected result

Succeeded

### Actual result

Error:
```none
Module 'Wyomind_Framework':
Module 'Wyomind_CronScheduler':
There is an error in /magento/vendor/magento/framework/Setup/Patch/PatchApplier.php at line: 255
get_class(): Argument #1 ($object) must be of type object, string given#0 /magento/vendor/magento/framework/Setup/Patch/PatchApplier.php(255): get_class('Wyomind\\CronSch...')
#1 /magento/setup/src/Magento/Setup/Model/Installer.php(1095): Magento\Framework\Setup\Patch\PatchApplier->applySchemaPatch('Wyomind_CronSch...')
#2 /magento/setup/src/Magento/Setup/Model/Installer.php(898): Magento\Setup\Model\Installer->handleDBSchemaData(Object(Magento\Setup\Module\Setup), 'schema', Array)
#3 /magento/setup/src/Magento/Setup/Console/Command/UpgradeCommand.php(146): Magento\Setup\Model\Installer->installSchema(Array)
#4 /magento/vendor/symfony/console/Command/Command.php(298): Magento\Setup\Console\Command\UpgradeCommand->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#5 /magento/vendor/symfony/console/Application.php(1040): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#6 /magento/vendor/symfony/console/Application.php(301): Symfony\Component\Console\Application->doRunCommand(Object(Magento\Setup\Console\Command\UpgradeCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#7 /magento/vendor/magento/framework/Console/Cli.php(118): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#8 /magento/vendor/symfony/console/Application.php(171): Magento\Framework\Console\Cli->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))╷
```

If we look at the code, we will see the error in it:

<img width="968" alt="image" src="https://github.com/magento/magento2/assets/1829515/7753fefb-2445-4031-9823-d11e65194216">

The `$schemaPatch` variable is changing its type from string to an object. But if an exception occurred before it becomes an object (i.e. in `isSkipableBySchemaSetupVersion` or `fixPatch` methods) the string value will be passed to the `get_class` function and cause the error.

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by engcom-Dash** [comment]
Hi @speller ,

verified the issue in 2.4-develop instance and the issue is not reproducible.

Kindly refer the below screenshots:
<img width="1512" alt="Screenshot 2023-06-07 at 9 02 37 AM" src="https://github.com/magento/magento2/assets/60198592/2cef6d4c-7759-42ab-be07-cc52c132cb6d">

Here we can run the setup upgrade command without any issues.

Kindly let me know if the issue still reproducible.

Thanks

**Comment by hostep** [comment]
@engcom-Dash: that's not the correct way to check this bug, the correct way is by reading the code posted by @speller in his initial post.

If `$schemaPatch` is of type `string` [on line 230](https://github.com/magento/magento2/blob/827073af53a60dd2cf0a4a83dfe2976cb7f97a15/lib/internal/Magento/Framework/Setup/Patch/PatchApplier.php#L230) and an exception gets thrown **before** it's turned into an object [on line 242](https://github.com/magento/magento2/blob/827073af53a60dd2cf0a4a83dfe2976cb7f97a15/lib/internal/Magento/Framework/Setup/Patch/PatchApplier.php#L242), then in the exception catching code, [line 255](https://github.com/magento/magento2/blob/827073af53a60dd2cf0a4a83dfe2976cb7f97a15/lib/internal/Magento/Framework/Setup/Patch/PatchApplier.php#L255) will fail because it's still a `string`.

Does that make it more clear?

A potential solution could be:
- on line 242, use a different variable to keep the object around, like `$schemaPatchObject` and use that variable from there on forth
- on line 255, use the original `$schemaPatch` variable without a `get_class` call

**Comment by hostep** [comment]
Closed without a comment explaining why?
Most likely fixed by [AC-9027::get_class(): Argument #1 () must be of type object, string given in magento/framework/Setup/Patch/PatchApplier](https://github.com/magento/magento2/commit/32189da9f4da70c87f655e463a45c4526ef05504)
